### PR TITLE
skeleton for basic AliceO2 simulation-point class

### PR DESCRIPTION
**Pull request by sawenzel**
Drafting a configurable hits/point class for detector simulation. This class could be a good prototype for our needs and requires less memory than FairMCPoint.

I am also trying to advocate here, the use of strong types for cartesian points and vectors. I am using in this example the types from ROOT::GenVector.

Ideally we should  define these geometry types (such as Point3D, Vector3D, Transform3D) centrally for consistent use throughout AliceO2.

### CODE REVIEW

File: Detectors/Base/include/DetectorsBase/BaseHits.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#include "FairMultiLinkedData_Interface.h"
+#include "Math/GenVector/DisplacementVector3D.h"
+#include "Math/GenVector/PositionVector3D.h"
+#include "Math/SVector.h"
+
+template <typename T>
+using Point3D = ROOT::Math::PositionVector3D<ROOT::Math::Cartesian3D<T>, ROOT::Math::DefaultCoordinateSystemTag>;
+template <typename T>
+using Vector3D = ROOT::Math::DisplacementVector3D<ROOT::Math::Cartesian3D<T>, ROOT::Math::DefaultCoordinateSystemTag>;
+
```
Comment (sawenzel) [inline Detectors/Base/include/DetectorsBase/BaseHits.h]: I would advocate for a global definition of these types and consistent use in our interfaces.

File: Detectors/Base/include/DetectorsBase/BaseHits.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// MC truth handling
+template <typename T>
+class BasicXYZEHit : public FairMultiLinkedData_Interface
+{
+  int mDetectorID; // the detector/sensor id
+  int mTrackID;    // this might be part of link
+  T mTime;         // time of flight
+  Point3D<T> mPos; // cartesian position of Hit
+  T mELoss;
+
```
Comment (shahor02) [inline Detectors/Base/include/DetectorsBase/BaseHits.h]: I think at least the mEloss and mTime can be fixed to float, in any case, their precision and operations on them are not related to that of mPos.
Also mDetectorID can be fixed to short (sorry for telling you opposite before, detectors will have internal enumeration of their sensors and we will have < 25k sensors).

Cheers,
Ruben

Comment (sawenzel) [inline Detectors/Base/include/DetectorsBase/BaseHits.h]: Thanks Ruben. Will adapt the class accordingly.

**Comment by sawenzel** [comment]
@ktf:  Can you give your opinion on Point3D + Vector3D ... they should most likely be in central framework directories.

**Comment by sawenzel** [comment]
@shahor02: Are you ok with the modified version?

**Comment by shahor02** [comment]
Hi @sawenzel , ok for energy/time members. For the coordinates, I have a question: the typical use-case for the coord.vector in aliroot is:

    double xyzGlo[3] = {hit.GetX(),hit.GetY(),hit.GetZ()}, xyzLoc[3];
    TGeoHMatrix *mat = ...; // take from manager or compose on the fly
    mat->MasterToLocal(xyzGlo,xyzLoc);

In case of hits, the coordinate always remains in the global frame, but for other classes requiring coordinate (e.g. cluster) we may need to transform the Point3D data member itself to other frame.

We better have the same types for all coordinates. Could you show the corresponding snippet in O2?


What would be the equivalent operation in O2?

**Comment by sawenzel** [comment]
@shahor02:

We could still use 
```c++
double xyzGlo[3] = {hit.GetX(),hit.GetY(),hit.GetZ()}, xyzLoc[3];
TGeoHMatrix *mat = ...; // take from manager or compose on the fly
mat->MasterToLocal(xyzGlo,xyzLoc);
```

However, I would be in favor to go to
```c++
ROOT::Math::Transform3D *trans = ...; // take from manager or compose on the fly
auto local = trans(hit.GetPos());
```
which is better typed; avoids pointers (and hence aliasing problems) + can easily mix double transformations + float points;


Similar things can be done for magnetic field lookup etc.

**Comment by shahor02** [comment]
Hi @sawenzel : trying to implement the (mis)alignment object noticed that the ROOT::Math::Transform3D has no templated operators for Point3D<T>, instead it operates on XYZPoint (or XYZVector) which is a double realization of Poin3D:  https://root.cern.ch/root/html528/ROOT__Math__XYZPoint.html.
Are you planning to ask root people to update it?

**Pull request merged**